### PR TITLE
SALTO-1388: added unresolved references core validator

### PR DESCRIPTION
**Pull request by alonstern**
Added to core a change validator to remove elements with unresolved references from the plan.

---
_Release Notes_: 
Core:
Elements with unresolved references will be removed from the plan with an indicative message

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/adapter-api/src/values.ts [hunk 1]
```diff
@@ -78,6 +78,11 @@ export class StaticFile {
   }
 }
 
+export class UnresolvedReference {
```
Comment (ori-moisis) [inline packages/adapter-api/src/values.ts:81]: why is this part of the adapter api?
the adapter should never get an unresolved reference

File: packages/core/src/core/plan/change_validators/unresolved_references.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { values, collections } from '@salto-io/lowerdash'
+
+const { awu } = collections.asynciterable
+
+const getUnresolvedReferences = async (element: Element): Promise<ElemID[]> => {
+  const unresolvedReferences: ElemID[] = []
+  await transformElement({
+    element,
+    strict: false,
+    runOnFields: true,
```
Comment (ori-moisis) [inline packages/core/src/core/plan/change_validators/unresolved_references.ts:27]: why do we need to run on fields? fields can't be references...

File: packages/core/test/core/plan/change_validators/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import { AdapterOperations, ElemID, ObjectType, ReferenceExpression, toChange, UnresolvedReference } from '@salto-io/adapter-api'
+import getChangeValidators from '../../../../src/core/plan/change_validators'
+
+describe('getChangeValidators', () => {
+  it('should call both the adapter change validators and the core change validators', async () => {
+    const adapterChangeValidator = jest.fn().mockResolvedValue([])
```
Comment (ori-moisis) [inline packages/core/test/core/plan/change_validators/index.test.ts:21]: use `mockFunction<ChangeValidator>()` to ensure the correct type

File: packages/core/test/core/plan/change_validators/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const changesValidators = getChangeValidators({
+      adapter: {
+        deployModifiers: {
+          changeValidator: adapterChangeValidator,
+        },
+      },
+    } as unknown as Record<string, AdapterOperations>)
```
Comment (ori-moisis) [inline packages/core/test/core/plan/change_validators/index.test.ts:28]: should not cast to `unknown` as this breaks the type coupling and could lead this test to be broken if and when the interface changes

File: packages/core/test/core/plan/change_validators/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    const type = new ObjectType({
+      elemID: new ElemID('adapter', 'type'),
+      annotations: {
+        value: new ReferenceExpression(new ElemID('adapter', 'someId'), new UnresolvedReference(new ElemID('adapter', 'someId'))),
+      },
+    })
+    const errors = await changesValidators.adapter([toChange({ after: type })])
+    expect(errors.length).toBe(1)
+    expect(errors[0].message).toBe('Element has unresolved references')
+    expect(adapterChangeValidator).toHaveBeenCalled()
```
Comment (ori-moisis) [inline packages/core/test/core/plan/change_validators/index.test.ts:39]: this does not test that the adapter change validator is actually used (only checks that it is called, but doesn't check that it gets relevant input or that the result from the adapter validator is returned)

File: packages/core/src/core/plan/change_validators/unresolved_references.ts (hunk 2 above)
Comment (alonstern) [reply packages/core/src/core/plan/change_validators/unresolved_references.ts:27]: Maybe I miss-understand what `runOnFields` do but annotation on fields can be references and this is required to iterate on those

Comment (ori-moisis) [reply packages/core/src/core/plan/change_validators/unresolved_references.ts:27]: you miss-understood what it does

File: packages/core/test/core/plan/change_validators/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        detailedMessage: '',
+        severity: 'Warning',
+      },
+    ])
+    const changesValidators = getChangeValidators({
+      adapter: {
+        deployModifiers: {
+          changeValidator: adapterChangeValidator,
+        },
+        fetch: jest.fn(),
```
Comment (ori-moisis) [inline packages/core/test/core/plan/change_validators/index.test.ts:36]:
```suggestion
        fetch: mockFunction<AdapterOperations['fetch']>(),
```

File: packages/core/test/core/plan/change_validators/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    const type = new ObjectType({
+      elemID: new ElemID('adapter', 'type'),
+      annotations: {
+        value: new ReferenceExpression(new ElemID('adapter', 'someId'), new expressions.UnresolvedReference(new ElemID('adapter', 'someId'))),
+      },
+    })
+    const changes = [toChange({ after: type })]
+    const errors = await changesValidators.adapter(changes)
+    expect(errors.length).toBe(2)
```
Comment (ori-moisis) [inline packages/core/test/core/plan/change_validators/index.test.ts:49]:
```suggestion
    expect(errors).toHaveLength(2)
```
this is important because when the test fails it would show the array and not just `1 != 2`

File: packages/workspace/src/expressions.ts [hunk 8]
```diff
@@ -20,6 +20,11 @@ import { collections, promises } from '@salto-io/lowerdash'
 
 const { mapValuesAsync } = promises.object
 
+export class UnresolvedReference {
```
Comment (ori-moisis) [inline packages/workspace/src/expressions.ts:23]: nit: why was this moved?

**Review by ori-moisis — APPROVED** [review]
